### ISSUE TITLE
Rule 932200, now inspecting Referer headers, matches any query string that contains spaces

### ISSUE THREAD

**Issue by RedXanadu** [coreruleset/coreruleset#3180]
### Description

Rule 932200 was recently updated to add the following items for inspection:

``REQUEST_HEADERS:Referer|REQUEST_HEADERS:User-Agent``

Unfortunately, with the _Referer_ header, this rule now causes false positives for the following scenario:

- referer header contains a URL, and
- URL includes a query string, and
- query string includes one or more space characters (`+`) **and/or** URL includes one or more space characters (`%20`).

For example, these headers both cause a false positive:

`Referer: http://www.example.com/page?param=test+test'`

`Referer: http://www.example.com/page%20test?param=test'`

This header does _not_ cause a false positive, as there are no encoded space characters:

`Referer: http://www.example.com/page?param=test_no_space_char'`

---

#### Cause

Rule 932200 has a regular expression looking for one of a number of optional patterns. This is the problematic pattern:

``/[^/]+?[*?`\x5c']``

`<forward slash><non-forward slash chars><question mark>` -> this will match URLs that feature a query string, e.g.

`/some/dir/page?foo=bar`

The presence of this chained rule:

``SecRule MATCHED_VAR "@rx \s" "t:none,t:urlDecodeUni``

means that a space character must also be present.

### How to reproduce the misbehavior (-> curl call)

Testing the CRS Docker container:

```
curl -v -o/dev/null localhost -A 'test' -H 'Referer: http://www.example.com/page?param=test+test'
```

=>

``
[Thu Mar 30 11:13:44.997007 2023] [security2:error] [pid 40:tid 140616113563392] [client 172.19.0.1:54008] [client 172.19.0.1] ModSecurity: Warning. Pattern match "\\\\s" at MATCHED_VAR. [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf"] [line "942"] [id "932200"] [msg "RCE Bypass Technique"] [data "Matched Data: /page? found within REQUEST_HEADERS:Referer: http://www.example.com/page?param=test test"] [severity "CRITICAL"] [ver "OWASP_CRS/4.0.0-rc1"] [tag "modsecurity"] [tag "application-multi"] [tag "language-multi"] [tag "platform-multi"] [tag "attack-rce"] [tag "paranoia-level/2"] [tag "OWASP_CRS"] [tag "capec/1000/152/248/88"] [tag "PCI/6.5.2"] [hostname "localhost"] [uri "/"] [unique_id "ZCVu6EXwaRSBQCgL9kHo9wAAAIw"], referer: http://www.example.com/page?param=test+test
``

### Your Environment

* CRS version (e.g., v3.3.4): v4.0.0-rc1
* Paranoia level setting (e.g. PL1) : PL 2
* ModSecurity version (e.g., 2.9.6): n/a
* Web Server and version or cloud provider / CDN (e.g., Apache httpd 2.4.54): Apache
* Operating System and version: n/a

### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by theseion** [comment]
I vote for simply removing `Referer` matching again. There's no quick fix here I can see, we would need to craft a new rule that targets the actual format of the `Referer` header and checks for URL related bypass techniques specifically.

**Comment by Xhoenix** [comment]
I currently using the following rule as an exclusion. It handles the above situation and another situation that my personal blog faced.

```
SecRule SERVER_NAME "@contains blog.jitendrapatro.me" \
     "id:1058,\
     phase:1,\
     pass,\
     nolog,\
     chain"
     SecRule REQUEST_HEADERS:Referer "@rx /\\w+(?:\.|'')\\w+\\?" \
        "ctl:ruleRemoveTargetById=932200;REQUEST_HEADERS:Referer"
```

**Comment by lifeforms** [comment]
I also think we should remove Referer again.

**Comment by Xhoenix** [comment]
If you're going to remove it, then atleast take into consideration what @theseion suggested. There are cases in which the Referer header can be exploited by attackers and we need to have separate rule targeting common bypass techniques.

**Comment by Xhoenix** [comment]
I thought this over and I think removing is totally not an solution. Rather, there should be a separate rule specifically targeting common bypasses.

**Comment by Xhoenix** [comment]
How did you get the Let's Encrypt FP? I've set Certbot to automatically update Certificates at the first of every month and have never faced any problems.

**Comment by emphazer** [comment]
@Xhoenix I installed the newest v4 release on several servers and they had all the same problem...

**Comment by theseion** [comment]
There's already a separate issue for this: #3163.

**Comment by Xhoenix** [comment]
@emphazer I think the problem is while installing new certificates and not while renewing them. Did the FP also shows up when renewing certs?

**Comment by Xhoenix** [comment]
@RedXanadu How about this:-
```/[^ a-zA-Z0-9._-]+?[*?`\x5c'])```

In the above regex if I've used the `/` then a payload like `cat /etc/passwd` would've been allowed. This above also excludes common command injection characters like ```&|;`$\``` and more. This also protects against all command injection payloads presently [known publicly](https://github.com/swisskyrepo/PayloadsAllTheThings/tree/master/Command%20Injection).

**For a basic test:-**

```sh
echo "$testing/t so-me.php?isanothertest" | grep -oP "/[ a-zA-Z0-9._/-]+"
```

Is there any reason this shouldn't be used? I'll be happy to open a PR if you want this. 🙂

**Comment by Xhoenix** [comment]
Okay, the hyphen character(`-`) allows payloads like `sh -c whoami`, but will most likely get blocked by other rules. Still, I think removing it would be a better choice.

**Comment by RedXanadu** [comment]
@Xhoenix Not sure if that would work, you'd have to test it. (I'm not working on a fix for this issue, just FYI.)

The outcome from the meeting discussion was to try and fix the existing regular expression. If that does not work then 'Referer' inspection for this case will be moved to paranoia level 3 as we cannot afford these false positives at PL 2.

**Comment by theseion** [comment]
I've opened a PR that addresses this issue: #3300. I was forced to create two additional rules. @Xhoenix, I would appreciated it if you'd also take a look at the PR.

**Comment by Xhoenix** [comment]
Small typos in rule 932200 regex. Updated regex can be viewed here:- https://regex101.com/r/w3kLqB/1

**Comment by theseion** [comment]
As far as I can tell, your version is identical to the one in this PR...?

**Comment by RedXanadu** [comment]
@Xhoenix There's [a CRS rule](https://coreruleset.org/docs/development/contribution_guidelines/#forward-slash-representation) that states that forward slashes are not escaped in regular expression patterns, in the interests of making patterns slightly easier to read.

@theseion I've suggested (via a PR to your PR) to also add tests to check against the original rule that was reported to have false positives. I've also asked a question in the PR review as I don't follow what part of the PR is doing.

**Comment by theseion** [comment]
@RedXanadu I can't find the PR that you say you've opened.

**Comment by theseion** [comment]
Found it. Since the PR has already been merged, I've created a new PR with your changes: #3309.
